### ISSUE TITLE
Cannot deserialize a 'Dictionary<String, String>' from BsonType 'String'

### ISSUE THREAD

**Issue by sgrassie** [sergeyzwezdin/Hangfire.Mongo#59]
When attempting to look at a particular job through the Dashboard, this exception is thrown. The job doesn't appear to be anything special. I suspect this problem will be related to Dashboard changes in Hangfire.Core

```
[FormatException: Cannot deserialize a 'Dictionary<String, String>' from BsonType 'String'.]
   MongoDB.Bson.Serialization.Serializers.DictionarySerializerBase`3.DeserializeValue(BsonDeserializationContext context, BsonDeserializationArgs args) +102
   MongoDB.Bson.Serialization.Serializers.ClassSerializerBase`1.Deserialize(BsonDeserializationContext context, BsonDeserializationArgs args) +161
   MongoDB.Bson.Serialization.Serializers.SerializerBase`1.MongoDB.Bson.Serialization.IBsonSerializer.Deserialize(BsonDeserializationContext context, BsonDeserializationArgs args) +14
   MongoDB.Driver.Linq.Translators.ProjectedObjectDeserializer.ReadDocument(BsonDeserializationContext context, String currentKey, String scopeKey, ProjectedObject currentObject) +249
   MongoDB.Driver.Linq.Translators.ProjectedObjectDeserializer.Deserialize(BsonDeserializationContext context, BsonDeserializationArgs args) +121
   MongoDB.Bson.Serialization.IBsonSerializerExtensions.Deserialize(IBsonSerializer`1 serializer, BsonDeserializationContext context) +97
   MongoDB.Bson.Serialization.Serializers.ProjectingDeserializer`2.Deserialize(BsonDeserializationContext context, BsonDeserializationArgs args) +76
   MongoDB.Bson.Serialization.IBsonSerializerExtensions.Deserialize(IBsonSerializer`1 serializer, BsonDeserializationContext context) +97
   MongoDB.Driver.Core.Operations.CursorBatchDeserializationHelper.DeserializeBatch(RawBsonArray batch, IBsonSerializer`1 documentSerializer, MessageEncoderSettings messageEncoderSettings) +665
   MongoDB.Driver.Core.Operations.FindCommandOperation`1.CreateCursorBatch(BsonDocument commandResult) +292
   MongoDB.Driver.Core.Operations.FindCommandOperation`1.CreateCursor(IChannelSourceHandle channelSource, BsonDocument commandResult, Boolean slaveOk) +157
   MongoDB.Driver.Core.Operations.FindCommandOperation`1.Execute(IReadBinding binding, CancellationToken cancellationToken) +394
   MongoDB.Driver.Core.Operations.FindOperation`1.Execute(IReadBinding binding, CancellationToken cancellationToken) +306
   MongoDB.Driver.OperationExecutor.ExecuteReadOperation(IReadBinding binding, IReadOperation`1 operation, CancellationToken cancellationToken) +67
   MongoDB.Driver.MongoCollectionImpl`1.ExecuteReadOperation(IReadOperation`1 operation, ReadPreference readPreference, CancellationToken cancellationToken) +153
   MongoDB.Driver.MongoCollectionImpl`1.ExecuteReadOperation(IReadOperation`1 operation, CancellationToken cancellationToken) +88
   MongoDB.Driver.MongoCollectionImpl`1.FindSync(FilterDefinition`1 filter, FindOptions`2 options, CancellationToken cancellationToken) +226
   MongoDB.Driver.FindFluent`2.ToCursor(CancellationToken cancellationToken) +125
   MongoDB.Driver.IAsyncCursorSourceExtensions.ToList(IAsyncCursorSource`1 source, CancellationToken cancellationToken) +84
   Hangfire.Mongo.<>c__DisplayClass5_0.<JobDetails>b__0(HangfireDbContext connection) in C:\projects\hangfire-mongo\artifacts\sources\Hangfire.Mongo\MongoMonitoringApi.cs:100
   Hangfire.Dashboard.Pages.JobDetailsPage.Execute() +198
   Hangfire.Dashboard.RazorPage.TransformText(String body) +31
   Hangfire.Dashboard.RazorPageDispatcher.Dispatch(DashboardContext context) +86
   Hangfire.Dashboard.<>c__DisplayClass1_2.<UseHangfireDashboard>b__1(IDictionary`2 env) +468
   Microsoft.Owin.Mapping.<Invoke>d__0.MoveNext() +462
   System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) +13891908
   System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) +61
   Microsoft.Owin.Host.SystemWeb.IntegratedPipeline.<RunApp>d__5.MoveNext() +202
   System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) +13891908
   System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) +61
   Microsoft.Owin.Mapping.<Invoke>d__0.MoveNext() +870
   System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) +13891908
   System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) +61
   Microsoft.Owin.Host.SystemWeb.IntegratedPipeline.<RunApp>d__5.MoveNext() +202
   System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) +13891908
   System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) +61
   Microsoft.Owin.Host.SystemWeb.IntegratedPipeline.<DoFinalWork>d__2.MoveNext() +193
   Microsoft.Owin.Host.SystemWeb.IntegratedPipeline.StageAsyncResult.End(IAsyncResult ar) +96
   System.Web.AsyncEventExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute() +363
   System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously) +137
```

**Comment by gottscj** [comment]
Hi,

Thanks, sorry for the late reply.

I'll take a look.

**Comment by pavlo-p** [comment]
The problem is that StateDto.Data was changed from String to Dictionary<string, string> and now if you have a portion of data with old string representation then serialization fails. If you want I can implement custom serializer for it.

**Comment by gottscj** [comment]
That would be great. :)

Thanks!

**Comment by pavlo-p** [comment]
Pitfall of custom serialization is that it could slow down performance. Maybe it is better to use migration script for it and change all strings to objects like this:

`
db.hangfire.state.find({Data: {$type: "string"}}).noCursorTimeout().forEach(
	function(state){
	  if(state.Data){
	    try {
	      var jsonData = JSON.parse(state.Data);
	      db.hangfire.state.updateOne(
	      	{_id: state._id}, 
	      	{$set: {"Data": jsonData}});
		} catch (e) {
   			print(e);
		}
	  }
	});

`

**Comment by pavlo-p** [comment]
But I don't know how you will include it in release.

**Comment by sgrassie** [comment]
I would suggest the migration script option. I can test it next week on a
large system.

On Thu, 6 Apr 2017, 16:46 Pavlo Podopryhora, <notifications@github.com>
wrote:

Pitfall of custom serialization is that it could slow down performance.
Maybe it is better to use migration script for it and change all strings to
objects like this:

`
db.hangfire.state.find({Data: {$type: "string"}}).noCursorTimeout().forEach(
function(state){
if(state.Data){
try {
var jsonData = JSON.parse(state.Data);
db.hangfire.state.updateOne(
{_id: state._id},
{$set: {"Data": jsonData}});
} catch (e) {
print(e);
}
}
});

`

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/sergeyzwezdin/Hangfire.Mongo/issues/59#issuecomment-292216577>,
or mute the thread

**Comment by pavlo-p** [comment]
I just tested it on test db and it took 20 minutes to update 200k.

**Comment by pavlo-p** [comment]
@gottscj, what do you think about script approach?

**Comment by gottscj** [comment]
@pavlo-p,

Thanks! :) The best approach would be migrations. We would like to take the time to define a strategy for that.
For now I have pushed a branch where I implemented a simple serializer which checks for BsonType.String.

I would not expect a performance overhead as we, before, used NewtonSoft to deserialize the string to Dictionary<string, string>. Now were just doing it with the BsonSerializer. check it out.

We are struggling a little bit with the new project files in netstandard, also AppVeyor is not happy. We will fix it ASAP.
